src/omnipool/invariants.rs

```diff
@@ -0,0 +1,323 @@
+
+            let asset_out_state = asset_out.clone();
+            let asset_out_state = asset_out_state.delta_update(&state_changes.asset_out).unwrap();
+            assert_asset_invariant(&asset_out, &asset_out_state,  FixedU128::from((TOLERANCE, ONE)), "Buy update invariant - token out");
+        }
+    }
+}
+#[test]
+fn buy_update_invariants_no_fees_case() {
+    let asset_in = AssetReserveState {
+        reserve: 10000000000000000,
+        hub_reserve: 10000000000000000,
+        shares: 10000000000000000,
+        protocol_shares: 10000000000000000,
+        tvl: 10000000000000000,
+    };
+    let asset_out = AssetReserveState {
+        reserve: 10000000000000000,
+        hub_reserve: 89999999999999991,
+        shares: 10000000000000000,
+        protocol_shares: 10000000000000000,
+        tvl: 10000000000000000,
+    };
+    let amount = 1_000_000_000_000_000;
+
+    let result = calculate_buy_state_changes(
+        &asset_in,
+        &asset_out,
+        amount,
+        Permill::from_percent(0),
+        Permill::from_percent(0),
+        Balance::default(),
+    );
+
+    assert!(result.is_none()); // This fails because of not enough asset out in pool out
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1000))]
+    #[test]
+    fn buy_update_invariants_no_fees(asset_in in asset_state(), asset_out in asset_state(),
+        amount in trade_amount()
+    ) {
+        let result = calculate_buy_state_changes(&asset_in, &asset_out, amount,
+
+        Permill::from_percent(0),
+        Permill::from_percent(0),
+            Balance::default()
+        );
+
+        // ignore the invalid result
+        if let Some(state_changes) = result {
+            let asset_in_state = asset_in.clone();
+            let asset_in_state = asset_in_state.delta_update(&state_changes.asset_in).unwrap();
+            assert_asset_invariant(&asset_in, &asset_in_state,  FixedU128::from((TOLERANCE, ONE)), "Buy update invariant - token in");
+
+            let asset_out_state = asset_out.clone();
+            let asset_out_state = asset_out_state.delta_update(&state_changes.asset_out).unwrap();
+            assert_asset_invariant(&asset_out, &asset_out_state,  FixedU128::from((TOLERANCE, ONE)), "Buy update invariant - token out");
+        }
+    }
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1000))]
+    #[test]
+    fn add_liquidity_prices(asset in asset_state(),
+        amount in trade_amount(),
+        stable_asset in stable_asset_state()
+    ) {
+        let result = calculate_add_liquidity_state_changes(&asset,
+            amount,
+            stable_asset,
+            false
+        );
+
+        assert!(result.is_some());
+
+        let state_changes = result.unwrap();
+
+        let new_asset_state= asset.clone();
+        let new_asset_state = new_asset_state.delta_update(&state_changes.asset).unwrap();
+
+        // Price should not change
+        assert_eq_approx!(asset.price().unwrap(),
```

Comment:
I would extract it to a helper method called `assert_that_price_has_not_changed`, then also using it in the next test.

By doing so, we can also get rid of the comment.

The 3rd param of this approx method can be more generic, and if the test fails, we can easily get the corresponding reason if you do my proposed test renamings.